File: app/src/stores/dedup.ts
import { defineStore } from 'pinia'
import { computed, ref } from 'vue'
import type { UnlistenFn } from '@tauri-apps/api/event'
import {
  cancelDetectDuplicates as ipcCancel,
  isTauri,
  onDedupCancelled,
  onDedupComplete,
  onDedupError,
  onDedupProgress,
  scanDetectDuplicates as ipcStart,
  type DedupCandidate,
  type DuplicateGroup,
} from '@/api/tauri'
import { notify } from '@/lib/notify'
import { i18n } from '@/i18n'

const t = (key: string, params?: Record<string, unknown>) =>
  params ? i18n.global.t(key, params) : i18n.global.t(key)

/**
 * S14 重复文件检测 store。
 *
 * 设计原则:
 *
 * 1. **冷启动**:store 默认 idle,用户在 Scan 页 Tab 切到「重复文件」
 *    时按需触发,避免每次扫描完成自动跑(BLAKE3 大文件仍要几秒)。
 * 2. **解耦 scan**:输入是 scan store 当前的 results 派生候选,不读
 *    DB — 用户可以随时切换扫描根目录后再 rerun。
 * 3. **三段进度**:`size` / `head` / `full`,UI 据 stage 渲染不同
 *    文案,total + processed 喂进度条。
 * 4. **事件订阅幂等**:`ensureSubscribed` 自带 guard,Settings 页 +
 *    Scan 页 + 命令面板任意入口都可以安全调用。
 */
export type DedupPhase = 'idle' | 'size' | 'head' | 'full' | 'done' | 'cancelled' | 'error'

export const useDedupStore = defineStore('dedup', () => {
  const phase = ref<DedupPhase>('idle')
  const processed = ref(0)
  const total = ref(0)
  const groups = ref<DuplicateGroup[]>([])
  const totalWastedBytes = ref(0)
  const durationMs = ref(0)
  const errorMessage = ref<string | null>(null)
  /** 上一次 detect 的候选总数,UI 在 "size" 阶段还没分组前先展示个估值 */
  const candidatesCount = ref(0)

  let unlistenProgress: UnlistenFn | null = null
  let unlistenComplete: UnlistenFn | null = null
  let unlistenCancelled: UnlistenFn | null = null
  let unlistenError: UnlistenFn | null = null

  const running = computed(
    () => phase.value === 'size' || phase.value === 'head' || phase.value === 'full',
  )

  const progressPercent = computed(() => {
    if (total.value === 0) return 0
    return Math.min(100, Math.round((processed.value / total.value) * 100))
  })

  async function ensureSubscribed() {
    if (unlistenProgress && unlistenComplete && unlistenCancelled && unlistenError) return
    if (!isTauri()) return
    unlistenProgress = await onDedupProgress(p => {
      // 后端 stage 是 "size" / "head" / "full",直接映射到 phase。
      // 重要:size 阶段 emit 时 phase 可能仍是 idle(首个事件触发),
      // 这是 OK 的 — 走 if 后才决定 phase。
      if (p.stage === 'size') phase.value = 'size'
      else if (p.stage === 'head') phase.value = 'head'
      else if (p.stage === 'full') phase.value = 'full'
      processed.value = p.processed
      total.value = p.total
    })
    unlistenComplete = await onDedupComplete(p => {
      console.info('[dedup] complete', {
        groups: p.groups.length,
        wasted: p.totalWastedBytes,
        ms: p.durationMs,
      })
      groups.value = p.groups
      totalWastedBytes.value = p.totalWastedBytes
      durationMs.value = p.durationMs
      phase.value = 'done'
    })
    unlistenCancelled = await onDedupCancelled(() => {
      console.info('[dedup] cancelled by user')
      phase.value = 'cancelled'
    })
    unlistenError = await onDedupError(p => {
      console.error('[dedup] error', p.message)
      errorMessage.value = p.message
      phase.value = 'error'
      notify.error(t('dedup.errorTitle'), p.message)
    })
  }

  /**
   * 启动 detect。candidates 应当来自 scan store 的 results(扁平化为
   * id + path + sizeBytes)。重复触发(running 状态下)会被拒绝。
   */
  async function detect(candidates: DedupCandidate[], minSizeBytes?: number) {
    if (running.value) {
      console.warn('[dedup] detect ignored — already running phase', phase.value)
      return
    }
    if (!isTauri()) {
      errorMessage.value = t('dedup.browserMode')
      phase.value = 'error'
      notify.error(errorMessage.value)
      return
    }
    if (candidates.length === 0) {
      errorMessage.value = t('dedup.noCandidates')
      phase.value = 'error'
      return
    }

    await ensureSubscribed()
    reset(false)
    candidatesCount.value = candidates.length
    phase.value = 'size'

    try {
      await ipcStart({ candidates, minSizeBytes })
      console.info('[dedup] backend accepted scan_detect_duplicates', {
        candidates: candidates.length,
        minSizeBytes,
      })
    } catch (e) {
      const msg = String(e)
      console.error('[dedup] start IPC rejected', e)
      errorMessage.value = msg
      phase.value = 'error'
      notify.error(t('dedup.startFailedTitle'), msg)
    }
  }

  async function cancel() {
    if (!isTauri()) return
    if (!running.value) return
    try {
      await ipcCancel()
    } catch {
      /* noop */
    }
  }

  /**
   * 重置临时计数。`keepResults=true` 时保留 groups 数据(用户切走 Tab
   * 再回来不应丢失结果),仅清进度状态 — 但默认 false,detect 重启时
   * 是要把上次结果连同进度一起清掉。
   */
  function reset(keepResults = true) {
    phase.value = 'idle'
    processed.value = 0
    total.value = 0
    errorMessage.value = null
    if (!keepResults) {
      groups.value = []
      totalWastedBytes.value = 0
      durationMs.value = 0
      candidatesCount.value = 0
    }
  }

  return {
    phase,
    processed,
    total,
    groups,
    totalWastedBytes,
    durationMs,
    errorMessage,
    candidatesCount,
    running,
    progressPercent,
    ensureSubscribed,
    detect,
    cancel,
    reset,
  }
})

File: app/src-tauri/src/commands/dedup.rs
//! `scan_detect_duplicates` / `cancel_detect_duplicates` — S14 重复文件
//! 检测的 IPC 入口。
//!
//! 设计原则:
//!
//! 1. **解耦扫描**:输入是前端 `scan.results` 派生的候选列表(path +
//!    size + id),不依赖 DB 状态。这样用户可以在 dedup 完成后单独 rerun
//!    而不必重扫整盘;也方便测试。
//! 2. **后台线程 + 事件**:与 `start_scan` 一致,IPC 同步返回 Ok 后启动
//!    后台 hash 任务,通过 `dedup:progress` / `dedup:complete` /
//!    `dedup:cancelled` / `dedup:error` 流式上报。
//! 3. **互斥执行**:同一时刻只允许一个 dedup 任务在跑(rayon 全局线程
//!    池本身也不喜欢嵌套),`dedup_running` 是互斥标志。

use std::sync::atomic::Ordering;
use std::time::Instant;

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter, State};

use crate::scanner::dedup::{
    self, DedupCandidate, DedupError, DedupProgress, DedupStage, DuplicateGroup,
    DEFAULT_MIN_SIZE_BYTES,
};
use crate::state::ScanState;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DedupArgs {
    /// 来自前端 scan.results 的扁平候选列表。每项至少含 path + sizeBytes
    /// + id(便于 UI 在结果展示阶段映射回原扫描行)。
    candidates: Vec<DedupCandidate>,
    /// 单个文件最小入选大小,默认 1 MB。前端通常硬编码或暴露在「设置 →
    /// 扫描」里。低于该值的文件被直接丢弃,避免空文件 / 缩略图浪费时间。
    #[serde(default)]
    min_size_bytes: Option<u64>,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
struct DedupProgressPayload {
    /// dedup 阶段标签:`size` / `head` / `full`。前端按此切换文案 + 进
    /// 度条颜色。`size` 阶段近乎瞬完成,主要让 UI 立即把"候选 N 个"展
    /// 示出来,避免 head 阶段大文件时长时间空白。
    stage: &'static str,
    processed: u64,
    total: u64,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
struct DedupCompletePayload {
    groups: Vec<DuplicateGroup>,
    total_wasted_bytes: u64,
    duration_ms: u128,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
struct DedupErrorPayload {
    message: String,
}

#[tauri::command]
pub fn scan_detect_duplicates(
    args: DedupArgs,
    app: AppHandle,
    state: State<'_, ScanState>,
) -> Result<(), String> {
    if state.dedup_running.swap(true, Ordering::SeqCst) {
        eprintln!("[diskmind] scan_detect_duplicates rejected: previous run still in flight");
        return Err(crate::i18n::i18n("scanner.dedup.error.already_running"));
    }
    state.dedup_cancel.store(false, Ordering::SeqCst);

    let candidates_count = args.candidates.len();
    let min_size = args.min_size_bytes.unwrap_or(DEFAULT_MIN_SIZE_BYTES);
    eprintln!(
        "[diskmind] scan_detect_duplicates accepted: {} candidates, min_size={}",
        candidates_count, min_size
    );

    let cancel = state.dedup_cancel.clone();
    let running = state.dedup_running.clone();
    let app_handle = app.clone();

    std::thread::spawn(move || {
        let started = Instant::now();
        let progress_handle = app_handle.clone();

        let result = dedup::detect_duplicates(
            args.candidates,
            min_size,
            cancel,
            move |progress: DedupProgress| {
                let stage = match progress.stage {
                    DedupStage::SizeGrouping => "size",
                    DedupStage::HeadHashing => "head",
                    DedupStage::FullHashing => "full",
                };
                let _ = progress_handle.emit(
                    "dedup:progress",
                    DedupProgressPayload {
                        stage,
                        processed: progress.processed,
                        total: progress.total,
                    },
                );
            },
        );

        running.store(false, Ordering::SeqCst);

        match result {
            Ok(groups) => {
                let total_wasted: u64 = groups.iter().map(|g| g.wasted_bytes).sum();
                let duration_ms = started.elapsed().as_millis();
                let groups_count_log = groups.len();
                let payload = DedupCompletePayload {
                    groups,
                    total_wasted_bytes: total_wasted,
                    duration_ms,
                };
                match app_handle.emit("dedup:complete", payload) {
                    Ok(()) => eprintln!(
                        "[diskmind] dedup:complete emitted (groups={}, wasted={}, ms={})",
                        groups_count_log, total_wasted, duration_ms
                    ),
                    Err(e) => eprintln!("[diskmind] dedup:complete emit failed: {e}"),
                }
            }
            Err(DedupError::Cancelled) => {
                eprintln!("[diskmind] dedup cancelled by user");
                let _ = app_handle.emit("dedup:cancelled", ());
            }
        }
    });

    Ok(())
}

#[tauri::command]
pub fn cancel_detect_duplicates(state: State<'_, ScanState>) -> Result<(), String> {
    state.dedup_cancel.store(true, Ordering::SeqCst);
    Ok(())
}

// 用于 lib.rs 注册阶段 — 把 `DedupErrorPayload` 列出来,避免 dead_code
// 警告。当前 dedup 路径只有 cancel + done 两种结局,error 走 emit 已经
// 在 thread 内部处理(实际不会发到 DedupErrorPayload,这里保留作为未
// 来扩展锚点 — IO 错误聚合后的 fatal 路径)。
#[allow(dead_code)]
fn _hold_error_payload_type(p: DedupErrorPayload) -> DedupErrorPayload {
    p
}
